gensim/models/fasttext_inner.pyx

```diff
@@ -0,0 +1,544 @@
+    const np.uint32_t *subwords_index, const np.uint32_t subwords_len, 
+    const REAL_t alpha, REAL_t *work, REAL_t *l1, REAL_t *word_locks_vocab,
+    REAL_t *word_locks_ngrams) nogil:
+
+    cdef long long a, b
+    cdef np.uint32_t word2_index = subwords_index[0]
+    cdef long long row1 = word2_index * size, row2, sgn
+    cdef REAL_t f, g, f_dot, lprob
+
+    memset(work, 0, size * cython.sizeof(REAL_t))
+    memset(l1, 0, size * cython.sizeof(REAL_t))
+
+    scopy(&size, &syn0_vocab[row1], &ONE, l1, &ONE)
+    for d in range(1, subwords_len):
+        our_saxpy(&size, &ONEF, &syn0_ngrams[subwords_index[d] * size], &ONE, l1, &ONE)
+    cdef REAL_t norm_factor = ONEF / subwords_len
+    sscal(&size, &norm_factor, l1 , &ONE)
+
+    for b in range(codelen):
+        row2 = word_point[b] * size
+        f_dot = our_dot(&size, l1, &ONE, &syn1[row2], &ONE)
+        if f_dot <= -MAX_EXP or f_dot >= MAX_EXP:
+            continue
+        f = EXP_TABLE[<int>((f_dot + MAX_EXP) * (EXP_TABLE_SIZE / MAX_EXP / 2))]
+        g = (1 - word_code[b] - f) * alpha
+
+        our_saxpy(&size, &g, &syn1[row2], &ONE, work, &ONE)
+        our_saxpy(&size, &g, l1, &ONE, &syn1[row2], &ONE)
+
+    our_saxpy(&size, &word_locks_vocab[word2_index], work, &ONE, &syn0_vocab[row1], &ONE)
+    for d in range(1, subwords_len):
+        our_saxpy(&size, &word_locks_ngrams[subwords_index[d]], work, &ONE, &syn0_ngrams[subwords_index[d]*size], &ONE)
+
+
+cdef unsigned long long fast_sentence_cbow_neg(
+    const int negative, np.uint32_t *cum_table, unsigned long long cum_table_len, int codelens[MAX_SENTENCE_LEN],
+    REAL_t *neu1,  REAL_t *syn0_vocab, REAL_t *syn0_ngrams, REAL_t *syn1neg, const int size,
+    const np.uint32_t indexes[MAX_SENTENCE_LEN], const np.uint32_t *subwords_idx[MAX_SENTENCE_LEN],
+    const int subwords_idx_len[MAX_SENTENCE_LEN], const REAL_t alpha, REAL_t *work,
+    int i, int j, int k, int cbow_mean, unsigned long long next_random, REAL_t *word_locks_ngrams, REAL_t *word_locks_vocab) nogil:
+
+    cdef long long a
+    cdef long long row2
+    cdef unsigned long long modulo = 281474976710655ULL
+    cdef REAL_t f, g, count, inv_count = 1.0, label, log_e_f_dot, f_dot
+    cdef np.uint32_t target_index, word_index
+    cdef int d, m
+
+    word_index = indexes[i]
+
+    memset(neu1, 0, size * cython.sizeof(REAL_t))
+    count = <REAL_t>0.0
+    for m in range(j, k):
+        if m == i:
+            continue
+        else:
+            count += ONEF
+            our_saxpy(&size, &ONEF, &syn0_vocab[indexes[m] * size], &ONE, neu1, &ONE)
+            for d in range(subwords_idx_len[m]):
+                count += ONEF
+                our_saxpy(&size, &ONEF, &syn0_ngrams[subwords_idx[m][d] * size], &ONE, neu1, &ONE)
+
+    if count > (<REAL_t>0.5):
+        inv_count = ONEF/count
+    if cbow_mean:
+        sscal(&size, &inv_count, neu1, &ONE)  # (does this need BLAS-variants like saxpy?)
```

Comment:
I can see these comments and the TODOs are copied over fro `word2vec_inner`. Is now a good time to addressing them?